### PR TITLE
Buildenv update

### PR DESCRIPTION
**Pull request by craiglink**
-     add support for ESP32S3
-     add support for ESP32-S3-Matrix
-     update Build Environment to latest Ardunio-Espressif32 PlatformIO build 6.7.0 / 3.20017.0
-     update libraries include FastLED which supports Hardware SPI for ESP32

### CODE REVIEW

**Comment by craiglink** [comment]
@SteveEisner  is issue I was seeing this weekend is from another code change / optimization that is not present in the pull request

File: tools/WLED_ESP32_4MB_noOTA.csv [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Name,   Type, SubType, Offset,  Size, Flags
```
Comment (SteveEisner) [inline tools/WLED_ESP32_4MB_noOTA.csv:1]: wow. I don't know how to review this ;)

File: usermods/Tubes/debug.h [hunk 2]
```diff
@@ -71,8 +71,8 @@ class DebugController {
         knownIp[2],
         knownIp[3],
         freeMemory(),
-        LITTLEFS.usedBytes(),
-        LITTLEFS.totalBytes(),
+        WLED_FS.usedBytes(),
```
Comment (SteveEisner) [inline usermods/Tubes/debug.h]: we can rebase this out now

File: usermods/audioreactive/audio_reactive.h [hunk 3]
```diff
@@ -300,7 +300,7 @@ void FFTcode(void * parameter)
 #endif
 
 #ifdef UM_AUDIOREACTIVE_USE_NEW_FFT
-      FFT.majorPeak(FFT_MajorPeak, FFT_Magnitude);                // let the effects know which freq was most dominant
+      FFT.majorPeak(&FFT_MajorPeak, &FFT_Magnitude);                // let the effects know which freq was most dominant
```
Comment (SteveEisner) [inline usermods/audioreactive/audio_reactive.h]: we should contribute this upstream?  or I wonder if I merge the branch to a later version if they've found and fixed this

File: usermods/Tubes/Tubes.h [hunk 4]
```diff
@@ -38,11 +38,16 @@ class TubesUsermod : public Usermod {
 
   public:
     void setup() {
-      pinMode(MASTER_PIN, INPUT_PULLUP);
-      pinMode(LEGACY_PIN, INPUT_PULLUP);
-      if (digitalRead(MASTER_PIN) == LOW) {
+
+      if (pinManager.isPinOk(MASTER_PIN)) {
```
Comment (SteveEisner) [inline usermods/Tubes/Tubes.h:42]: this is great - I assume we need to fix code in master.h too, where it reads button pins etc

File: platformio.ini [hunk 5]
```diff
@@ -47,8 +47,9 @@ default_envs = nodemcuv2, esp8266_2m, esp01_1m_full, nodemcuv2_160, esp8266_2m_1
 src_dir  = ./wled00
 data_dir = ./wled00/data
 build_cache_dir = ~/.buildcache
```
Comment (SteveEisner) [inline platformio.ini:49]: I'll pull this into my branch, it's mostly unreviewed
I suspect we should just contribute this part upstream to get a good review on it
and so we don't have to continuously merge

**Review by SteveEisner — APPROVED** [review]

**Comment by SteveEisner** [comment]
Looks like it fails on two builds due to VSPI being missing. I'll try bringing it local to understand why that's not available, unless you know of an obvious reason based on your changes!

Builds failing: **lolin_s2_mini**, **esp32c3dev**    (but for instance **esp32s3dev_8MB** passes!)

**Comment by craiglink** [comment]
I had not looked at the s2 and c3 builds. They definitely are more limited chipsets. They might not support hardware spi, so I might just need to look more at the platform io inheritance.

What is your current thought about being able to contribute bat to WLED ?

On Jun 30, 2024, at 3:21 PM, Steve Eisner ***@***.***> wrote:

﻿

Looks like it fails on two builds due to VSPI being missing. I'll try bringing it local to understand why that's not available, unless you know of an obvious reason based on your changes!

**Comment by SteveEisner** [comment]
I'd like to be able to contribute back parts of this. Not immediately but eventually.

There are some simple things - like the bug you found in audioreactive - that seem like easy candidates.  Changing the platformio.ini will be harder, but having their review on it will be useful and I think they'd appreciate the help upgrading to later FastLED and so on.

**Comment by SteveEisner** [comment]
For now, I'd just like to get the build process fixed so that I can merge the PR.  I could merge it broken but then it'll continue to break for all future PRs, might as well investigate to see what's wrong!

**Comment by SteveEisner** [comment]
Seems related to: https://github.com/FastLED/FastLED/issues/1615

File: usermods/audioreactive/audio_reactive.h (hunk 3 above)
Comment (SteveEisner) [reply usermods/audioreactive/audio_reactive.h]: Looks like they've done something similar already: https://github.com/SteveEisner/WLEDtubes/commit/ecc9443677e1c69291138e8bf8f54d524614a144

Comment (SteveEisner) [reply usermods/audioreactive/audio_reactive.h]: I've just merged this into the WLEDtubes/main branch, so you should be able to rebase/merge your PR branch and remove this fix

File: platformio.ini [hunk 6]
```diff
@@ -286,15 +262,15 @@ lib_deps =
...
-platform_packages =
+platform = ${esp32.platform}
+platform_packages = ${esp32.platform_packages}
 build_flags = -g
-  -DARDUINO_ARCH_ESP32
-  -DARDUINO_ARCH_ESP32C3
-  -DCONFIG_IDF_TARGET_ESP32C3=1
+  -D ARDUINO_ARCH_ESP32
+  -D ARDUINO_ARCH_ESP32C3
+  -D FASTLED_ALL_PINS_HARDWARE_SPI
```
Comment (SteveEisner) [inline platformio.ini:270]: and `-D FASTLED_ESP32_SPI_BUS=HSPI` in this case

Comment (SteveEisner) [reply platformio.ini:270]: this plus `-D FASTLED_ESP32_SPI_BUS=FSPI` for the `lolin_s2_mini` target gets past the build errors.

**Pull request merged**